README.md
# Vault CA
[![Build Status](https://travis-ci.org/crisidev/vault-ca.svg?branch=master)](https://travis-ci.org/crisidev/vault-ca) [![codecov](https://codecov.io/gh/crisidev/vault-ca/branch/master/graph/badge.svg)](https://codecov.io/gh/crisidev/vault-ca) [![Code Climate](https://codeclimate.com/github/crisidev/vault-ca/badges/gpa.svg)](https://codeclimate.com/github/crisidev/vault-ca) [![Issue Count](https://codeclimate.com/github/crisidev/vault-ca/badges/issue_count.svg)](https://codeclimate.com/github/crisidev/vault-ca) [![Documentation Status](https://readthedocs.org/projects/vault-ca/badge/?version=latest)](http://vault-ca.readthedocs.io/en/latest/?badge=latest)

Set of tools to create your own CA and manage certificates using hashicorp Vault.

## Install
```bash
apt-get install libffi-dev libssl-dev
pip install vault-ca
```

## Documentation
[https://vault-ca.readthedocs.io/en/latest/index.html](https://vault-ca.readthedocs.io/en/latest/index.html)

## How to
### Initialize the Vault
Vault need to be initialized without SSL certificates and than switched.
Once initialized and configured, fetch-ssl-cert can be used to retrieve
Vault's own certificate/key pair and TLS can be switched on.

Inside the the `example` folder you can find two different Vault server
configurations for this purpose

```bash
❯❯❯ ~ vault server -config example/vault.hcl.init

❯❯❯ ~ export VAULT_ADDR="http://127.0.0.1:8200"

❯❯❯ ~ vault init
Unseal Key 1: SmMc4xXT2oq6d7uJSnMMnuRKLh7EighJUeH4Kh/28naG
Unseal Key 2: ovsPuWWd8dWVBqb+TPtqqQRjGPBNCor3rp1QLYYnhNZJ
Unseal Key 3: XuASk26YBqQo9+cvi17Me5o7PWsQwVlL2hM8G1bjOyIs
Unseal Key 4: NwhxHrkByuVfP4GURwDc/tZ1qPP6bVsmP2hUIQj+nvBN
Unseal Key 5: 0Kfq2jVlE2Db8Nj5mIXVhQTRONYz/0qZ1zm1BObUO4nx
Initial Root Token: 087efd94-3680-0ff4-25b7-a92177cd3c46

Vault initialized with 5 keys and a key threshold of 3. Please
securely distribute the above keys. When the vault is re-sealed,
restarted, or stopped, you must provide at least 3 of these keys
to unseal it again.

Vault does not store the master key. Without at least 3 keys,
your vault will remain permanently sealed.

❯❯❯ ~ vault unseal SmMc4xXT2oq6d7uJSnMMnuRKLh7EighJUeH4Kh/28naG
Sealed: true
Key Shares: 5
Key Threshold: 3
Unseal Progress: 1
Unseal Nonce: 4ba54ac9-d60b-4078-05da-285ef05bf303
❯❯❯ ~ vault unseal ovsPuWWd8dWVBqb+TPtqqQRjGPBNCor3rp1QLYYnhNZJ
Sealed: true
Key Shares: 5
Key Threshold: 3
Unseal Progress: 2
Unseal Nonce: 4ba54ac9-d60b-4078-05da-285ef05bf303
❯❯❯ ~ vault unseal XuASk26YBqQo9+cvi17Me5o7PWsQwVlL2hM8G1bjOyIs
Sealed: false
Key Shares: 5
Key Threshold: 3
Unseal Progress: 0
Unseal Nonce:

❯❯❯ ~ vault status
Sealed: false
Key Shares: 5
Key Threshold: 3
Unseal Progress: 0
Unseal Nonce:
Version: 0.7.2
Cluster Name: test-cluster
Cluster ID: 8f948916-8a3d-b220-f0b4-7c5bb74dc5a6

High-Availability Enabled: false
```

### Intialize certificate authority
After initializing and unsealing the Vault, you need to create the certificate authority.

One of the script installed by **vault_ca**, `create-vault-ca`, will do than for you.

Default parameters:

* The script creates a certificate authority valid for 5 years.
* Certificates have a maximum TTL of 5 years and a default one of 30 days.
* Authorization tokens have a maximum TTL of 5 years and a default one of 1 year.

All the parameters above can be configured through command line arguments. See `create-vault-ca --help` for details.

NOTE: set `--vault-token` option using `Initial Root Token` from above.

```bash
❯❯❯ ~ create-vault-ca --domain=test.org --component=test --vault-addr="http://127.0.0.1:8200" --vault-token=087efd94-3680-0ff4-25b7-a92177cd3c46
Are you sure this script have not been already run on this vault? It can break / override configs [y|N] y
Successfully mounted 'pki' at 'pki/test.org'!
Successfully tuned mount 'pki/test.org'!
Key          	Value
---          	-----
certificate  	-----BEGIN CERTIFICATE-----
MIIC9jCCAd6gAwIBAgIUKybrEs7kUvRSgepQuImxonoznCwwDQYJKoZIhvcNAQEL
BQAwEzERMA8GA1UEAxMIdGVzdC5vcmcwHhcNMTcwNTE0MTcyNDMwWhcNMjIwNTEz
MTcyNTAwWjATMREwDwYDVQQDEwh0ZXN0Lm9yZzCCASIwDQYJKoZIhvcNAQEBBQAD
ggEPADCCAQoCggEBALeVV8zUKwhYJz0eVg+6rWfCPz+GdxxIXeiChULAU+zHWvDf
Jxye9JrcTdc/XUI0ZSw33F2JEjLkDasdchfL4ESRbUTdnJj1kYW6KEF9X3rhL/AM
hdX+EqUQ9yvXRlvcSyGObVD7ayRUcG2IDpCLRuFW5bkw+MxvSjyzIf6+W3bs5DVz
mFKqRv5Y3ycsuzc8CiDjxj/1LZWvBfqUFf8jePO0bzL3kw7uViZA4fJ23wPLqTyq
IRX52ODZFC3SeyF600lerCLGY4Bgol8YtZwjsx+MxpPnszlkitxT2wjAghfPTV0W
8BebQi4D+CN4A4C6joyGZrdagzsUF3LoeGoDFh8CAwEAAaNCMEAwDgYDVR0PAQH/
BAQDAgEGMA8GA1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYEFHovOm1ZHkAsmCGKn+u3
QmlCrriZMA0GCSqGSIb3DQEBCwUAA4IBAQBgDE//Tktbe16VSSrqP8MNYGtMG+jZ
PV2Ao7FCrgSPwjBQHPXR1fh+g4MMG4S9iI8QtXIz49/ZYXfXPU6LPq8W/zrIlf/g
7PODdOo/w7LA7CBHG6ceQtRXHuaMJvJ8EybTQ4vc7LK2FMdEZbBQQkQfnCunR8bz
oACVTooX2DkSPHCM24XSBsmMsHxImEYrjzsr0RyU+R9Tq+rdhjoEyUzQxklIecTq
8D1yfIrgIfYfT1qf6n2bEb+xIfk47v8yXlIUS3KLDadUtqybHIzsbSKEwiQse7rF
AOoUPGoZMSJAr52y0SW2QE8mJoGyX0HeqeX2ocrKw3WvwXF1oHpOB6Au
-----END CERTIFICATE-----
expiration   	1652462700
issuing_ca   	-----BEGIN CERTIFICATE-----
MIIC9jCCAd6gAwIBAgIUKybrEs7kUvRSgepQuImxonoznCwwDQYJKoZIhvcNAQEL
BQAwEzERMA8GA1UEAxMIdGVzdC5vcmcwHhcNMTcwNTE0MTcyNDMwWhcNMjIwNTEz
MTcyNTAwWjATMREwDwYDVQQDEwh0ZXN0Lm9yZzCCASIwDQYJKoZIhvcNAQEBBQAD
ggEPADCCAQoCggEBALeVV8zUKwhYJz0eVg+6rWfCPz+GdxxIXeiChULAU+zHWvDf
Jxye9JrcTdc/XUI0ZSw33F2JEjLkDasdchfL4ESRbUTdnJj1kYW6KEF9X3rhL/AM
hdX+EqUQ9yvXRlvcSyGObVD7ayRUcG2IDpCLRuFW5bkw+MxvSjyzIf6+W3bs5DVz
mFKqRv5Y3ycsuzc8CiDjxj/1LZWvBfqUFf8jePO0bzL3kw7uViZA4fJ23wPLqTyq
IRX52ODZFC3SeyF600lerCLGY4Bgol8YtZwjsx+MxpPnszlkitxT2wjAghfPTV0W
8BebQi4D+CN4A4C6joyGZrdagzsUF3LoeGoDFh8CAwEAAaNCMEAwDgYDVR0PAQH/
BAQDAgEGMA8GA1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYEFHovOm1ZHkAsmCGKn+u3
QmlCrriZMA0GCSqGSIb3DQEBCwUAA4IBAQBgDE//Tktbe16VSSrqP8MNYGtMG+jZ
PV2Ao7FCrgSPwjBQHPXR1fh+g4MMG4S9iI8QtXIz49/ZYXfXPU6LPq8W/zrIlf/g
7PODdOo/w7LA7CBHG6ceQtRXHuaMJvJ8EybTQ4vc7LK2FMdEZbBQQkQfnCunR8bz
oACVTooX2DkSPHCM24XSBsmMsHxImEYrjzsr0RyU+R9Tq+rdhjoEyUzQxklIecTq
8D1yfIrgIfYfT1qf6n2bEb+xIfk47v8yXlIUS3KLDadUtqybHIzsbSKEwiQse7rF
AOoUPGoZMSJAr52y0SW2QE8mJoGyX0HeqeX2ocrKw3WvwXF1oHpOB6Au
-----END CERTIFICATE-----
serial_number	2b:26:eb:12:ce:e4:52:f4:52:81:ea:50:b8:89:b1:a2:7a:33:9c:2c

Success! Data written to: pki/test.org/roles/cert
Policy 'pki/test.org/cert' written.
Success! Data written to: auth/token/roles/services
Success! Data written to: auth/token/roles/users

Generating services token (REMEMBER TO SAVE IT)
Key            	Value
---            	-----
token          	9a6b0ee1-c159-3710-dc3f-7641a5ef9222
token_accessor 	3730bbec-d8ae-eac0-cc26-c4fb6efa3e0a
token_duration 	8760h0m0s
token_renewable	true
token_policies 	[default pki/test.org/cert]

Have you saved the services token above? [y|N] y

Generating users token (REMEMBER TO SAVE IT)
Key            	Value
---            	-----
token          	3f2ebc35-1793-dc9e-f8da-f6dd6d081ca2
token_accessor 	7fcf4113-4918-54a1-d24c-8533b3b10e53
token_duration 	8760h0m0s
token_renewable	true
token_policies 	[default pki/test.org/cert]

Have you saved the users token above? [y|N] y

To boostrap the CA use "fetch-ssl-cert -c test -n <common_name> -t <services_token> -b -o <output_dir>"
To create / renew a certificate use "fetch-ssl-cert -c test -n <common_name> -t <services_token> -o <output_dir>"
```

From now on, you can use one of the two tokens created during the CA setup:

* services: used by automated services to fetch certificate/key pairs.
* users: used by humans to fetch certificate/key pairs.

### Secure Vault with TLS
Once the CA is setup, Vault itself need to get its own certificate/key pair and the CA need to be downloaded and bootstrapped into the system.

I am assuming you are on Debian and you have `update-ca-certificates` available.

NOTE: from now on the token used is `services` from above.

```bash
❯❯❯ fetch-ssl-cert -c test -n vault.test.org -d test.org -i 127.0.0.1 -t 9a6b0ee1-c159-3710-dc3f-7641a5ef9222 -b -A http://127.0.0.1:8200 -D
2017-05-14 18:52:58,443 __init__.py:44 DEBUG:vault address is `http://127.0.0.1:8200`
2017-05-14 18:52:58,443 __init__.py:174 DEBUG:request url is `http://127.0.0.1:8200/v1/pki/test.org/issue/cert`
2017-05-14 18:52:58,443 __init__.py:131 DEBUG:requesting new cert / key part for CA domain: `test.org`, component: `test`, common_name: `vault.test.org`, ip_sans: `127.0.0.1`, alt_names: `None`, ttl: `8760h`
2017-05-14 18:52:58,450 connectionpool.py:207 DEBUG:Starting new HTTP connection (1): 127.0.0.1
2017-05-14 18:52:58,590 connectionpool.py:395 DEBUG:http://127.0.0.1:8200 "PUT /v1/pki/test.org/issue/cert HTTP/1.1" 200 None
2017-05-14 18:52:58,591 __init__.py:59 DEBUG:directory `/usr/local/share/ca-certificates/test.org` already exists, skipping creation
2017-05-14 18:52:58,591 __init__.py:106 DEBUG:writing certificate for vault.test.org on /usr/local/share/ca-certificates/test.org/test-vault.test.org.pem
2017-05-14 18:52:58,591 __init__.py:109 DEBUG:writing private key for vault.test.org on /usr/local/share/ca-certificates/test.org/test-vault.test.org.key
2017-05-14 18:52:58,591 __init__.py:114 DEBUG:writing CA on /usr/local/share/ca-certificates/test.org/test.crt
```

Now that the certificate/key pair and the CA are saved on the disk, we need to update the system certificate authorities

```bash
❯❯❯ update-ca-certificates --fresh
Updating certificates in /etc/ssl/certs... 1 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d....done.
```

Stop vault and restart it using the other configuration from `example` directory and repeat the unseal process.

```bash
❯❯❯ vault server -config example/vault.hcl.ssl
```

Now vault is fully secure and usable.

NOTE: remember to register a local DNS for `vault.test.org` or use the command line option to specify the Vault address.

### Fetch certificate/key pair
Let's fetch a certificate/key pair for a test domain.

```bash
fetch-ssl-cert -c test -n test.test.org -d test.org -t 9a6b0ee1-c159-3710-dc3f-7641a5ef9222 -A https://127.0.0.1:8200 -D
2017-05-14 18:52:58,443 __init__.py:44 DEBUG:vault address is `https://127.0.0.1:8200`
2017-05-14 18:52:58,443 __init__.py:174 DEBUG:request url is `https://127.0.0.1:8200/v1/pki/test.org/issue/cert`
2017-05-14 18:52:58,443 __init__.py:131 DEBUG:requesting new cert / key part for CA domain: `test.org`, component: `test`, common_name: `test.test.org`, ip_sans: `None`, alt_names: `None`, ttl: `8760h`
2017-05-14 18:52:58,450 connectionpool.py:207 DEBUG:Starting new HTTPS connection (1): 127.0.0.1
2017-05-14 18:52:58,590 connectionpool.py:395 DEBUG:https://127.0.0.1:8200 "PUT /v1/pki/test.org/issue/cert HTTP/1.1" 200 None
2017-05-14 18:52:58,591 __init__.py:59 DEBUG:directory `/usr/local/share/ca-certificates/test.org` already exists, skipping creation
2017-05-14 18:52:58,591 __init__.py:106 DEBUG:writing certificate for test.test.org on /usr/local/share/ca-certificates/test.org/test-test.test.org.pem
2017-05-14 18:52:58,591 __init__.py:109 DEBUG:writing private key for vault.test.org on /usr/local/share/ca-certificates/test.org/test-test.test.org.key
```


debian/changelog
vault-ca (0.6-1) unstable; urgency=medium

  * New upstream version

 -- Matteo Bigoi <bigo@crisidev.org>  Wed, 05 Jul 2017 00:20:40 +0100

vault-ca (0.5-2.1) unstable; urgency=medium

  * Fix bootstrap argument parsing.

 -- Matteo Bigoi <bigo@crisidev.org>  Wed, 05 Jul 2017 00:02:23 +0100

vault-ca (0.5-2) unstable; urgency=low

  * update package support only python2 and python3

 -- Matteo Bigoi <bigo@crisidev.org>  Tue, 20 Jun 2017 19:53:31 +0000

vault-ca (0.5-1) unstable; urgency=low

  * source package automatically created by stdeb 0.8.2

 -- Matteo Bigoi <bigo@crisidev.org>  Tue, 06 Jun 2017 20:06:08 +0000


vault_ca/ca.py
import os
import logging
from datetime import datetime, timedelta

from OpenSSL.crypto import load_certificate, FILETYPE_PEM, Error


class VaultCertError(Exception):
    pass


class VaultCert(object):
    ASN1_GENERALIZETIME_FORMAT = "%Y%m%d%H%M%SZ"
    DEFAULT_CERTIFICATE_TTL = '8760h'  # 1 year
    DEFAULT_VALID_INTERVAL_DAYS = 1
    VAULT_ADDRESS = 'https://vault.{}:8002'
    CA_PATH = '/usr/local/share/ca-certificates/{}'

    def __init__(self, kwargs):
        self.component = kwargs['component']
        self.domain = kwargs['domain']
        self.vault_token = kwargs['vault_token']
        self.output_dir = kwargs.get('output_dir') or self.CA_PATH.format(self.domain)
        self.bootstrap_ca = kwargs.get('bootstrap_ca')
        self.ssl_verity = kwargs.get('ssl_verify')
        self.valid_interval = kwargs.get('valid_interval') or self.DEFAULT_VALID_INTERVAL_DAYS
        self.vault_address = kwargs.get('vault_address') or self.VAULT_ADDRESS.format(self.domain)
        self.ca_file = self.CA_PATH.format(self.domain)

    def _make_dirs(self):
        try:
            os.makedirs(self.output_dir)
        except OSError:
            logging.debug('directory %s already exists, skipping creation', self.output_dir)
            pass

    def _parse_asn1_generalizedtime(self, timestamp):
        try:
            datetime.strptime(timestamp, self.ASN1_GENERALIZETIME_FORMAT)
        except ValueError as e:
            logging.error("unable to parse timestamp %s into ASN.1 GENERALIZEDTIME: %s", timestamp, e.msg)
            raise VaultCertError("unable to parse certificate expire date")

    def _load_certificate(self, certificate_path):
        with open(certificate_path, 'r') as pem:
            try:
                certificate = load_certificate(FILETYPE_PEM, pem.read())
            except Error as e:
                logging.error("unable to load certificate %s: %s", certificate_path, e.msg)
                raise VaultCertError("unable to load certificate {}".format(certificate_path))
            else:
                return certificate

    def _is_certificate_valid(self, certificate_path):
        if os.path.exists(certificate_path):
            certificate = self._load_certificate(certificate_path)
            valid_not_after = self._parse_asn1_generalizedtime(certificate.get_notAfter.decode('utf-8'))
            tomorrow = datetime.now() + timedelta(days=self.valid_interval)
            if (tomorrow - valid_not_after) < timedelta(days=self.valid_interval):
                logging.debug(
                    "certificate %s is expiring in less than %d days, renewing required", certificate_path,
                    self.valid_interval
                )
                return False
            else:
                logging.debug(
                    "certificate %s is expiring in more than %d days, skipping renewal", certificate_path,
                    self.valid_interval
                )
                return True
        else:
            logging.debug("certificate %s does not exist on disk, fetching required", certificate_path)
            return False

    def _write_files(self, common_name, cert_data, priv_key_data, ca_data):
        cert_file = os.path.join(self.output_dir, '{}-{}.pem'.format(self.component, common_name))
        priv_key_file = os.path.join(self.output_dir, '{}-{}.key'.format(self.component, common_name))
        with open(cert_file, 'w') as cert, open(priv_key_file, 'w') as priv_key:
            logging.debug("writing certificate for %s on %s", common_name, cert_file)
            cert.write(cert_data)

            logging.debug("writing private key for %s on %s", common_name, priv_key_file)
            priv_key.write(priv_key_data)

            if self.bootstrap_ca:
                with open(self.ca_file, 'w') as ca:
                    logging.debug("writing CA on %s", self.ca_file)
                    ca.write(ca_data)

    def fetch(self, common_name, ip_sans=None, alt_names=None, ttl=None):
        pass


vault_ca/__init__.py
import os
import json
import logging
from datetime import datetime, timedelta

import requests
from OpenSSL.crypto import load_certificate, FILETYPE_PEM, Error

VERSION = 0.6

# fix incompatibility between python 3.4 and 3.5+ json implementation
if not hasattr(json, 'JSONDecodeError'):  # pragma: nocover
    json.JSONDecodeError = ValueError


class VaultCAError(Exception):
    """
    VaultCA custom exception
    """
    pass


class VaultCA(object):
    """
    Object to handle fetching ot certificate/key pairs and CA.
    """
    ASN1_GENERALIZEDTIME_FORMAT = "%Y%m%d%H%M%SZ"
    DEFAULT_CERTIFICATE_TTL = '8760h'  # 1 year
    DEFAULT_VALID_INTERVAL_DAYS = 1
    VAULT_ADDRESS = "https://vault.{}:8200"
    VAULT_PATH = "v1/pki/{}/issue/cert"
    CA_PATH = "/usr/local/share/ca-certificates/{}"
    MANDATORY_ARGS = ('component', 'domain', 'vault_token')

    def __init__(self, kwargs):
        """
        Initialize object and its attributes.

        :param kwargs: arguments dictionary
        :type kwargs: dict
        """
        self._validate_args(kwargs)
        self._manage_args(kwargs)

    def _manage_args(self, kwargs):
        """
        Setup object attributes for later usage.

        :param kwargs: arguments dictionary
        :type kwargs: dict
        """

        self.component = kwargs['component']
        self.domain = kwargs['domain']
        self.vault_token = kwargs['vault_token']
        self.output_dir = kwargs.get('output_dir') or self.CA_PATH.format(self.domain)
        self.bootstrap_ca = kwargs.get('bootstrap_ca')
        if self.bootstrap_ca:
            self.ssl_verify = False
        else:
            self.ssl_verify = kwargs.get('ssl_verify')
        self.valid_interval = kwargs.get('valid_interval') or self.DEFAULT_VALID_INTERVAL_DAYS
        self.vault_address = kwargs.get('vault_address') or self.VAULT_ADDRESS.format(self.domain)
        self.ca_path = kwargs.get('ca_path') or self.CA_PATH.format(self.domain)
        logging.debug('vault address is `%s`', self.vault_address)

    def _validate_args(self, kwargs):
        """
        Validate that mandatory arguments are present.

        :param kwargs: arguments dictionary
        :type kwargs: dict

        :raises VaultCAError: if there are missing mandatory arguments
        """

        missing_args = []
        for arg in self.MANDATORY_ARGS:
            if arg not in kwargs.keys():
                logging.error("missing mandatory init argument `%s`", arg)
                missing_args.append(arg)
        if missing_args:
            raise VaultCAError("missing mandatory init arguments `%s`", ', '.join(missing_args))

    def _make_dirs(self, directory):
        """
        Create directory if not present.

        :param directory: directory to create
        :type directory: str
        """
        if not os.path.exists(directory):
            os.makedirs(directory)

    def _parse_asn1_generalizedtime(self, timestamp):
        """
        Parse a timestamp coming from PyOpenSSL X509 object into datetime.

        :param timestamp: ASN.1 GENERALIZEDTIME timestamp
        :type timestamp: str

        :return: parsed timestamp
        :rtype: datetime.datetime

        :raises VaultCAError: if the format is not parsable
        """
        try:
            return datetime.strptime(timestamp, self.ASN1_GENERALIZEDTIME_FORMAT)
        except ValueError as e:
            logging.error("unable to parse timestamp `%s` into ASN.1 GENERALIZEDTIME: %s", timestamp, e)
            raise VaultCAError("unable to parse certificate expire date")

    def _load_certificate(self, certificate_path):
        """
        Load certificate or CA from disk.

        :param certificate_path: certificate path on disk
        :type certificate_path: str

        :return: the loaded certificate
        :rtype: OpenSSL.crypto.X509

        :raises VaultCAError: if the certificate is not loadable
        """
        with open(certificate_path, 'r') as pem:
            try:
                certificate = load_certificate(FILETYPE_PEM, pem.read())
            except Error as e:
                logging.error("unable to load certificate `%s`: %s", certificate_path, e)
                raise VaultCAError("unable to load certificate `{}`".format(certificate_path))
            else:
                return certificate

    def _is_certificate_valid(self, certificate_path):
        """
        Check if a certificate or CA is still valid.

        The check is done comparing the notValidAfter date of the certificate with a day in the future,
        based on `self.valid_interval` parameters.

        If the certificate is due to expire less than `self.valid_interval` days, it is marked it to be renewed.

        :param certificate_path: certificate path on disk
        :type certificate_path: str

        :return: certificate is valid or not
        :rtype: bool
        """
        if os.path.exists(certificate_path):
            certificate = self._load_certificate(certificate_path)
            valid_not_after = self._parse_asn1_generalizedtime(certificate.get_notAfter().decode('utf-8'))
            if (valid_not_after - datetime.now()) < timedelta(days=self.valid_interval):
                logging.debug(
                    "certificate `%s` is expiring in less than %d days, renewing required", certificate_path,
                    self.valid_interval
                )
                return False
            else:
                logging.debug(
                    "certificate `%s` is expiring in more than %d days, skipping renewal", certificate_path,
                    self.valid_interval
                )
                return True
        else:
            logging.debug("certificate `%s` does not exist on disk, fetching required", certificate_path)
            return False

    def _write_files(self, common_name, cert_data, cert_file, priv_key_data, priv_key_file, ca_data, ca_file):
        """
        Store certificate, private key and CA on disk, if the ones already found on disk are not valid anymore.

        :param common_name: common name for the certificate / key pair
        :type common_name: str
        :param cert_data: certificate data
        :type cert_data: str
        :param cert_file: certificate path on disk
        :type cert_file: str
        :param priv_key_data: private key data
        :type priv_key_data: str
        :param priv_key_file: private key path on disk
        :type priv_key_file: str
        :param ca_data: CA data
        :type ca_data: str
        :param ca_file: CA path on disk
        :type ca_file: str
        """
        if not self._is_certificate_valid(cert_file):
            with open(cert_file, 'w') as cert, open(priv_key_file, 'w') as priv_key:
                logging.debug("writing certificate for %s on %s", common_name, cert_file)
                cert.write(cert_data)

                logging.debug("writing private key for %s on %s", common_name, priv_key_file)
                priv_key.write(priv_key_data)

        if self.bootstrap_ca and not self._is_certificate_valid(ca_file):
            with open(ca_file, 'w') as ca:
                logging.debug("writing CA on %s", ca_file)
                ca.write(ca_data)

    def _prepare_json_data(self, common_name, ip_sans=None, alt_names=None, ttl=None):
        """
        Prepare the json payload to be sent to Vault with the new certificate request.

        :param common_name: common name for the certificate / key pair
        :type common_name: str
        :param ip_sans: list of IP for the current certificate, comma separated
        :type ip_sans: str
        :param alt_names: list of alternative names for the current certificate, comma separated
        :type alt_names: str
        :param ttl: TTL for the certificate / key pair
        :type ttl: str

        :return: json encoded payload
        :rtype: str
        """
        if not ttl:
            ttl = self.DEFAULT_CERTIFICATE_TTL

        data = {'common_name': common_name, 'ttl': ttl}

        if ip_sans:
            data['ip_sans'] = ip_sans

        if alt_names:
            data['alt_names'] = alt_names

        logging.debug(
            "requesting new cert / key part for CA domain: `%s`, component: `%s`, common_name: `%s`, ip_sans: `%s`, "
            "alt_names: `%s`, ttl: `%s`", self.domain, self.component, common_name, ip_sans, alt_names, ttl
        )

        return json.dumps(data)

    def _analise_response(self, response):
        """
        Analise response from Vault to find errors and extract the json response payload.

        :type reponse: vault HTTP response
        :type response: requests.models.Response

        :return: json response payload
        :rtype: dict

        :raises VaultCAError: if HTTP code is not ok or there are errors
        """
        if response.ok:
            try:
                payload = response.json()
            except json.JSONDecodeError as e:
                logging.error("error decoding json response: %s", e)
                raise VaultCAError("error decoding json response")
            else:
                errors = payload.get('errors')
                if errors:
                    logging.error("vault returned errors generating cert / key pair: %s", " ".join(errors))
                    raise VaultCAError("vault returned errors generating cert / key pair")
                else:
                    return payload
        else:
            logging.error("vault returned HTTP code `%s`", response.status_code)
            raise VaultCAError("vault returned HTTP error")

    def _extract_certificates(self, payload):
        """
        Extract certificate, private key and CA from Vault response.

        :param payload: json response payload
        :type payload: dict
        :return: certificate, private key and CA
        :rtype: tuple

        :raises VaultCAError: if certificate or private key or CA are missing
        """

        cert_data = payload.get('data', {}).get('certificate')
        if not cert_data:
            logging.error("vault payload is missing certificate data")
            raise VaultCAError("vault payload is missing certificate data")

        priv_key_data = payload.get('data', {}).get('private_key')
        if not priv_key_data:
            logging.error("vault payload is missing private key data")
            raise VaultCAError("vault payload is missing private key data")

        ca_data = payload.get('data', {}).get('issuing_ca')
        if not ca_data:
            logging.error("vault payload is missing CA data")
            raise VaultCAError("vault payload is missing CA data")

        return cert_data, priv_key_data, ca_data

    def fetch(self, common_name, ip_sans=None, alt_names=None, ttl=None):
        """
        Fetch new certificate / key pair from Vault.

        If attribute `self.bootstrap_ca` is set to True, also the CA is fetched.

        Fetched object are written on disk.

        :param common_name: common name for the certificate / key pair
        :type common_name: str
        :param ip_sans: list of IP for the current certificate, comma separated
        :type ip_sans: str
        :param alt_names: list of alternative names for the current certificate, comma separated
        :type alt_names: str
        :param ttl: TTL for the certificate / key pair

        :raises VaultCAError: if the request return errors
        """
        self._make_dirs(self.output_dir)
        self._make_dirs(self.ca_path)

        cert_file = os.path.join(self.output_dir, '{}-{}.pem'.format(self.component, common_name))
        priv_key_file = os.path.join(self.output_dir, '{}-{}.key'.format(self.component, common_name))
        ca_file = os.path.join(self.ca_path, "{}.crt".format(self.component))

        if not self._is_certificate_valid(cert_file) or not self._is_certificate_valid(ca_file):
            url = "{}/{}".format(self.vault_address, self.VAULT_PATH.format(self.domain))
            logging.debug("request url is `%s`", url)

            headers = {'X-Vault-Token': self.vault_token}
            data = self._prepare_json_data(common_name, ip_sans=ip_sans, alt_names=alt_names, ttl=ttl)

            try:
                response = requests.put(url, data=data, headers=headers, verify=self.ssl_verify)
            except requests.exceptions.RequestException as e:
                logging.error("exception connecting to vault endpoint: %s", e)
                raise VaultCAError("exception connecting to vault endpoint")
            else:
                payload = self._analise_response(response)

            cert_data, priv_key_data, ca_data = self._extract_certificates(payload)
            self._write_files(common_name, cert_data, cert_file, priv_key_data, priv_key_file, ca_data, ca_file)
